File: medium/39-combination-sum.ts
/**
 * Given an array of distinct integers and a target, return a list of all
 * combinations that sum up to the target. Numbers may be used more than once.
 *
 * @param candidates - Array of distinct integers
 * @param target - Target integer
 *
 * @returns - Array of arrays containing integer combinations summing to target
 */

function combinationSum(candidates: number[], target: number): number[][] {
  const combinations: number[][] = [];

  function backtrack(start: number, combination: number[], remainder: number) {
    if (remainder === 0) {
      // Push a copy to combinations since we will continue to manipulate array
      combinations.push([...combination]);
      return;
    }

    if (remainder < 0) {
      return;
    }

    for (let i = start; i < candidates.length; i++) {
      combination.push(candidates[i]);
      backtrack(i, combination, remainder - candidates[i]);

      // Backtrack once the branches from this node are finished
      combination.pop();
    }
  }

  backtrack(0, [], target);

  return combinations;
}

File: python/39-combination-sum.py
from typing import List

def combination_sum(candidates: List[int], target: int) -> List[List[int]]:
    """
    Given a list of candidate integers and a target, returns a list containing
    all distinct combinations of candidates that sum to the target.

    e.g. combination_sum([2, 3, 6, 7], 7) -> [[2, 2, 3], [7]]
    """

    combinations = []

    # i is start index in candidates to consider
    # curr is current combination of candidates
    def dfs(i: int, curr: List[int], remainder: int):
        # found valid combination
        if remainder == 0:
            combinations.append(curr.copy())
            return

        # overshot target or no more candidates to consider
        if i >= len(candidates) or remainder < 0:
            return

        # fully explore down tree that includes this value
        curr.append(candidates[i])
        dfs(i, curr, remainder - candidates[i])
        curr.pop()

        # explore down next tree that doesn't include this value
        dfs(i + 1, curr, remainder)

    dfs(0, [], target)
    return combinations

if __name__ == "__main__":
    assert combination_sum([2, 3, 6, 7], 7) == [[2, 2, 3], [7]]
    assert combination_sum([2, 3, 5], 8) == [[2, 2, 2, 2], [2, 3, 3], [3, 5]]
    assert combination_sum([2], 1) == []

    print("Passed")